image_processor.py

```diff
@@ -0,0 +1,198 @@
+import camera
+import segment
+import pickle as pickle
+import numpy as np
+import cv2
+import Color as c
+import jit_functions as jit_f
+
+
+class Object():
+    def __init__(self, x = -1, y = -1, size = -1, distance = -1, exists = False):
+        self.x = x
+        self.y = y
+        self.size = size
+        self.distance = distance
+        self.exists = exists
+
+    def __str__(self) -> str:
+        return "[Object: x={}; y={}; size={}; distance={}; exists={}]".format(self.x, self.y, self.size, self.distance, self.exists)
+
+    def __repr__(self) -> str:
+        return "[Object: x={}; y={}; size={}; distance={}; exists={}]".format(self.x, self.y, self.size, self.distance, self.exists)
+
+
+# results object of image processing. contains coordinates of objects and frame data used for these results
+class ProcessedResults():
+
+    def __init__(self, 
+                balls=[], 
+                basket_b = Object(exists = False), 
+                basket_m = Object(exists = False), 
+                color_frame = [],
+                depth_frame = [],
+                fragmented = [],
+                debug_frame = []) -> None:
+
+
+        self.balls = balls
+        self.basket_b = basket_b
+        self.basket_m = basket_m
+        self.color_frame = color_frame
+        self.depth_frame = depth_frame
+        self.fragmented = fragmented
+
+        # can be used to illustrate things in a separate frame buffer
+        self.debug_frame = debug_frame
+
+
+#Main processor class. processes segmented information
+class ImageProcessor():
+    def __init__(self, camera, color_config = "colors/colors.pkl", debug = False):
+        self.camera = camera
+        self.kernel = np.ones((3, 3), np.uint8)
+
+        self.color_config = color_config
+        with open(self.color_config, 'rb') as conf:
+            self.colors_lookup = pickle.load(conf)
+            self.set_segmentation_table(self.colors_lookup)
+
+        self.fragmented	= np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+
+        self.t_balls = np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+        self.t_basket_b = np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+        self.t_basket_m = np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+
+        self.debug = debug
+        self.debug_frame = np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+
+    def set_segmentation_table(self, table):
+        segment.set_table(table)
+
+    def start(self):
+        self.camera.open()
+
+    def stop(self):
+        self.camera.close()
+
+
+    def analyze_balls(self, t_balls, fragments) -> list:
+        t_balls = cv2.dilate(t_balls, self.kernel, iterations=2)
```

Comment:
If you have difficulties seeing further away balls consider using dilate and erode to merge noisy ball readings

Reply:
will do

Reply:
Quite brave to run it for 2 iterations but if it works for you. You can tweak it. Additionally I would suggest 
to use a kernel of shape, then it emulates the balls approximate shape better since you are searching for round objects:
```[
[0, 1, 0],
[1, 1, 1],
[0, 1, 0],
]